### PR TITLE
feat(e2e): Adding E2E tests to test passing custom object serializer

### PR DESCRIPTION
**Pull request by vinagesh**
Adding E2E test to ensure we can pass a custom Object Serializer. This should be the same for any ADT API call so I have only added one test that calls the serialize and deseriazlize of custom Serializer.

### CODE REVIEW

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/TestObjectSerializer.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+using Azure.Core.Serialization;
+
+namespace Azure.DigitalTwins.Core.Tests
+{
+    /// <summary>
+    /// Custom serializer to pass to the DigitalTwins client.
+    /// </summary>
+    public class TestObjectSerializer : ObjectSerializer
+    {
+        private static JsonObjectSerializer _serializer = new JsonObjectSerializer();
```
Comment (timtay-microsoft) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/tests/TestObjectSerializer.cs:20]: Can you add a check that the expected methods in here are actually called? Just to make sure that when a user sets the custom serializer, that it actually gets used?

**Review by timtay-microsoft — APPROVED** [review]
One small comment, but otherwise looks good

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs [hunk 2]
```diff
@@ -121,5 +121,37 @@ public void DigitalTwins_TwinNotExist_ThrowsNotFoundException()
...
         }
+
+        [Test]
+        public async Task DigitalTwinOperationsWithCustomObjectSerializer_Succeeds()
+        {
+            // arrange
+
+            DigitalTwinsClientOptions options = new DigitalTwinsClientOptions
+            {
+                Serializer = new TestObjectSerializer()
```
Comment (drwill-ms) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs:132]: What does `TestObjectSerializer` do?

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/TestObjectSerializer.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+using System.Runtime.Serialization.Formatters.Binary;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Serialization;
+
+namespace Azure.DigitalTwins.Core.Tests
+{
+    /// <summary>
+    /// Custom serializer to pass to the DigitalTwins client.
```
Comment (drwill-ms) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/tests/TestObjectSerializer.cs:16]: What does it do? How do we know it was used?

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs (hunk 2 above)
Comment (drwill-ms) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs:132]: I see the impl below now. I'm curious what we're validating.

Comment (vinagesh) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs:132]: Nothing special, other than object to stream and stream to object using JsonObjectSerializer internally.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/TestObjectSerializer.cs (hunk 3 above)
Comment (vinagesh) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/tests/TestObjectSerializer.cs:16]: I does not do anything special, just uses the JsonObjectSerializer internally. I chose to do that as we just want to test passing of a custom serializer as options and not any logic of how a serializer is implemented.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/TestObjectSerializer.cs (hunk 1 above)
Comment (vinagesh) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/tests/TestObjectSerializer.cs:20]: Very good idea! Let me figure out a way to do that. Or i'll add some custom implementation to ensure that was called.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/TestObjectSerializer.cs (hunk 3 above)
Comment (vinagesh) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/tests/TestObjectSerializer.cs:16]: I tested with a breakpoint but I will add a field to ensure that we know it was used in the test. It cannot be accessed directly so I'll have to add something to the serializer.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs [hunk 4]
```diff
@@ -121,5 +121,43 @@ public void DigitalTwins_TwinNotExist_ThrowsNotFoundException()
...
+
+            // create room twin
+            BasicDigitalTwin roomTwin = TestAssetsHelper.GetRoomTwinPayload(roomModelId);
+            await client.CreateOrReplaceDigitalTwinAsync(roomTwinId, roomTwin).ConfigureAwait(false);
+
+            roomTwin = await client.GetDigitalTwinAsync<BasicDigitalTwin>(roomTwinId).ConfigureAwait(false);
+
+            // assert
+            roomTwin.Should().NotBeNull();
+            serializer.isDeserializeCalled.Should().Be(true);
```
Comment (vinagesh) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs:159]: This should help determine that the custom serializer was called

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs (hunk 2 above)
Comment (vinagesh) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs]: We are trying to validate that we are able to pass a custom serializer and use it through the ADT client.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/TestObjectSerializer.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    /// </summary>
+    public class TestObjectSerializer : ObjectSerializer
+    {
+        private static JsonObjectSerializer _serializer = new JsonObjectSerializer();
+
+        // This field is used by the tests to confirm the function was called.
+        public bool isDeserializeCalled { get; set; }
```
Comment (drwill-ms) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/tests/TestObjectSerializer.cs:23]: `WasDeserializedCalled` - casing and verb tense

Better yet, you might track how many calls were made, and then assert > 0 or a specific number.

**Review by drwill-ms — APPROVED** [review]

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/TestObjectSerializer.cs (hunk 5 above)
Comment (vinagesh) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/tests/TestObjectSerializer.cs:23]: Sure that is a good idea too. I will change it to int and assert on number of calls.

Comment (vinagesh) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/tests/TestObjectSerializer.cs]: I tried out using an int. Everything works as expected in 'Live' mode. But the 'Record' mode is surely caching something, causing the expected number to continue to increase on every run. For now, I will leave it as bool for this check in but I'll investigate the issue with Azure Test Framework.

Comment (vinagesh) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/tests/TestObjectSerializer.cs]: Looks like I got used to java casing :P

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs [hunk 6]
```diff
@@ -121,5 +121,43 @@ public void DigitalTwins_TwinNotExist_ThrowsNotFoundException()
...
+
+            // create room twin
+            BasicDigitalTwin roomTwin = TestAssetsHelper.GetRoomTwinPayload(roomModelId);
+            await client.CreateOrReplaceDigitalTwinAsync(roomTwinId, roomTwin).ConfigureAwait(false);
+
+            roomTwin = await client.GetDigitalTwinAsync<BasicDigitalTwin>(roomTwinId).ConfigureAwait(false);
+
+            // assert
+            roomTwin.Should().NotBeNull();
+            serializer.WasDeserializeCalled.Should().Be(true);
```
Comment (drwill-ms) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs:159]: BeTrue

**Pull request merged**